[clang] Define ptrauth_string_discriminator builtin.

ahmedbougacha opened the pull request:
This exposes the ABI-stable hash function that allows computing a 16-bit discriminator from a constant string.

This allows manually matching the implicit string discriminators computed in the ABI (e.g., from mangled names for vtable pointer/entry signing), as well as enabling the use of interesting discriminators when manually annotating specific pointers with the __ptrauth qualifier.

ahmedbougacha committed:
[clang] Define ptrauth_string_discriminator builtin.

This exposes the ABI-stable hash function that allows computing a 16-bit
discriminator from a constant string.

This allows manually matching the implicit string discriminators
computed in the ABI (e.g., from mangled names for vtable pointer/entry
signing), as well as enabling the use of interesting discriminators when
manually annotating specific pointers with the __ptrauth qualifier.

Co-Authored-By: John McCall <rjmccall@apple.com>

ahmedbougacha committed:
Document ptrauth_string_discriminator in clang/docs/PointerAuthentication.

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
   return Call;
 }
 
+static ExprResult PointerAuthStringDiscriminator(Sema &S, CallExpr *call) {
```
Naming style of variables all over the function (use `PascalCase` https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly)

ahmedbougacha replied:
Clang originally had different conventions;  it usually gets (has gotten) updated over time for older code, I suppose now's the time for these changes.

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
+  Expr *arg = call->getArgs()[0]->IgnoreParenImpCasts();
+
+  // Operand must be an ordinary or UTF-8 string literal.
+  auto literal = dyn_cast<StringLiteral>(arg);
```
```suggestion
  const auto *Literal = dyn_cast<StringLiteral>(arg);
```

asl replied:
```suggestion
  const auto *Literal = dyn_cast<StringLiteral>(Arg);
```

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
+  if (checkPointerAuthEnabled(S, call)) return ExprError();
+
+  // We've already performed normal call type-checking.
+  Expr *arg = call->getArgs()[0]->IgnoreParenImpCasts();
```
```suggestion
  const Expr *arg = call->getArgs()[0]->IgnoreParenImpCasts();
```

asl replied:
```suggestion
  Expr *Arg = Call->getArg(0)->IgnoreParenImpCasts();
```

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
+  // Operand must be an ordinary or UTF-8 string literal.
+  auto literal = dyn_cast<StringLiteral>(arg);
+  if (!literal || literal->getCharByteWidth() != 1) {
+    S.Diag(arg->getExprLoc(), diag::err_ptrauth_string_not_literal)
```
This error message looks untested

asl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
   return Call;
 }
 
+static ExprResult PointerAuthStringDiscriminator(Sema &S, CallExpr *call) {
```
```suggestion
static ExprResult PointerAuthStringDiscriminator(Sema &S, CallExpr *Call) {
```

asl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
 }
 
+static ExprResult PointerAuthStringDiscriminator(Sema &S, CallExpr *call) {
+  if (checkPointerAuthEnabled(S, call)) return ExprError();
```
```suggestion
  if (checkPointerAuthEnabled(S, Call)) return ExprError();
```

asl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
+
+  // Operand must be an ordinary or UTF-8 string literal.
+  auto literal = dyn_cast<StringLiteral>(arg);
+  if (!literal || literal->getCharByteWidth() != 1) {
```
```suggestion
  if (!Literal || Literal->getCharByteWidth() != 1) {
```

asl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
+  auto literal = dyn_cast<StringLiteral>(arg);
+  if (!literal || literal->getCharByteWidth() != 1) {
+    S.Diag(arg->getExprLoc(), diag::err_ptrauth_string_not_literal)
+      << (literal ? 1 : 0)
```
```suggestion
      << (Literal ? 1 : 0)
```

asl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
+  if (!literal || literal->getCharByteWidth() != 1) {
+    S.Diag(arg->getExprLoc(), diag::err_ptrauth_string_not_literal)
+      << (literal ? 1 : 0)
+      << arg->getSourceRange();
```
```suggestion
      << Arg->getSourceRange();
```

asl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2156,6 +2156,24 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
+    return ExprError();
+  }
+
+  return call;
```
```suggestion
  return Call;
```

asl reviewed (commented):
See mostly code style nits above

kovdan01 reviewed (dismissed):
The changes themself in terms of functionality look OK to me - but I want someone else with deeper understanding of the context to take a look before this gets merged.

kovdan01 commented on clang/test/CodeGen/ptrauth-intrinsics.c:
It would be nice to also add a run line against `aarch64` or `aarch64-elf` generic triple in this and other tests (see code below). The logic tested is not apple-specific. The test would need a slight change since IR for ELF has functions marked as `dso_local` - you might probably want to use smth like `// CHECK-LABEL: define {{.*}}void @test_auth()` to match both MachO and ELF or, alternatively, use different check prefixes for function labels on MachO and ELF to avoid regex in match lines.

```
// RUN: %clang_cc1 -triple arm64-apple-ios -fptrauth-intrinsics -emit-llvm %s  -o - | FileCheck %s
// RUN: %clang_cc1 -triple aarch64-elf     -fptrauth-intrinsics -emit-llvm %s  -o - | FileCheck %s
```

This is probably a subject for another tiny PR since the RUN line is not touched here and the test already existed.

ahmedbougacha committed:
Address review feedback.

- add missing sema test
- also test ELF codegen for intrinsics
- various capitalization and type nits

ahmedbougacha committed:
Format.

kovdan01 reviewed (commented):
LGTM, but I want someone else with deeper understanding of the context to take a look before this gets merged.

asl commented on clang/lib/Headers/ptrauth.h:
```diff
@@ -15,6 +15,29 @@ typedef enum {
   ptrauth_key_asda = 2,
   ptrauth_key_asdb = 3,
+
+#ifdef __APPLE__
```
Looks like this is something from already merged commit. Worth rebasing?

asl commented on clang/test/Sema/ptrauth.c:
```diff
@@ -47,6 +47,17 @@ void test_blend_discriminator(int *dp, int (*fp)(int), int value) {
+void test_string_discriminator(const char *str) {
+  __builtin_ptrauth_string_discriminator(); // expected-error {{too few arguments}}
+  __builtin_ptrauth_string_discriminator(str, str); // expected-error {{too many arguments}}
+  (void) __builtin_ptrauth_string_discriminator("test string"); // no warning
```
Can we have `:"test string"` case here as well? As we're checking on the width of the literal character?

kovdan01 commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -328,6 +328,21 @@ be done in a single instruction with an immediate integer.
+``ptrauth_extra_data_t``.
+
+The result is always a constant expression.  The result value is never zero and
+always within range for both the ``__ptrauth`` qualifier and
```
Since we are talking about ranges here, it would be nice to have these ranges explicitly described in corresponding parts of docs. As far as I understand, different implementations are allowed to have different constraints on discriminator. It would be nice to somewhere describe actual figures for some specific implementation, e.g. arm64e. Please let me know if there is already a place for that which I've missed.

ahmedbougacha replied:
That's with `ptrauth_blend_discriminator`

kovdan01 reviewed (commented):
@ahmedbougacha After the latest force-push, the PR seems to contain SipHash-related changes from PRs 93902, 94393 and 94394. Could you please limit the changes only to those which are actually being intended to be merged as a part of this PR and resolve merge conflicts with the base branch?

I'd like to take one more look at this since there were updates since my latest feedback, but with different changes mixed, it's a bit messy.

asl commented:
@ahmedbougacha Will you please rebase

ahmedbougacha committed:
Adopt renamed getPointerAuthStableSipHash.

ahmedbougacha committed:
Address review feedback.

- tweaks to header/doc comments
- add test for wide string literal

asl reviewed (approved):
Thanks!
